### PR TITLE
Fixes #96: Improve string representation of sortedm2m relationships

### LINKED ISSUE #96: How to improve str() representation of relationships? Eg: when deleting from admin
I have two models: `Config` and `Template` ([django-netjsonconfig](https://github.com/openwisp/django-netjsonconfig)), `Config` has a sortedm2m relation to `Template`.

When deleting an object which has m2m relationships from admin interface, django says:

```
Are you sure?

    Are you sure you want to delete the selected template? All of the following objects and their related items will be deleted:

Summary

    - Config-template relationships: 6
    - Templates: 1

Objects

    - Template: [Generic-OpenWRT] Wireless settings
        - Config-template relationship: Config_templates object
        - Config-template relationship: Config_templates object
        - Config-template relationship: Config_templates object
        - Config-template relationship: Config_templates object
        - Config-template relationship: Config_templates object
        - Config-template relationship: Config_templates object
```

The repetition of `Config-template relationship: Config_templates object` conveys no information. I would like to display something more useful.

For example, each config has a name, let's say I'm deleting a template which was installed on two config objects named `Auditorium` and `Gym`

I've been able to accomplish this only through monkey patching:

```python
class Config(models.Model):
    # ...
    pass

def sortedm2m__str__(self):
    """
    Improves string representation of m2m relationship objects
    """
    return self.config.name

Config.templates.through.__str__ = sortedm2m__str__
```

But I'm wondering if there's a better way. Any suggestion?

### PR DESCRIPTION
**Pull request by rohithasrk**
Fixes #96 

This PR aims to improve the string representation of SortedM2M relationship when deleting one from an admin panel. I basically added a `base_class` parameter, a class which is inherited by the through-model of a sortedm2m relationship. This is basically so that the users of django-sortedm2m can add their own `__unicode__` method and improve the string representation.

For ex., in `example/testapp/models.py` we've passed a base_class argument when a relationship is being established. Check the image below,

sortedm2m

@gregmuellegger : Please review.

### CODE REVIEW

File: example/testapp/models.py [hunk 1]
```diff
@@ -9,10 +9,13 @@ class Car(models.Model):
     def __unicode__(self):
         return self.plate
```
Comment (nemesisdesign) [inline example/testapp/models.py]: follow [pep8](https://www.python.org/dev/peps/pep-0008/) conventions and add an additional blank line

File: example/testapp/models.py [hunk 2]
```diff
@@ -9,10 +9,13 @@ class Car(models.Model):
...
 
+class m2mprint:
+    def __unicode__(self):
+        return unicode(self.car) + " in " + unicode(self.parkingarea)
 
 class ParkingArea(models.Model):
     name = models.CharField(max_length=50)
-    cars = SortedManyToManyField(Car)
+    cars = SortedManyToManyField(Car, base_class=m2mprint)
```
Comment (nemesisdesign) [inline example/testapp/models.py]: follow [pep8](https://www.python.org/dev/peps/pep-0008/) conventions and add an additional blank line

File: example/testapp/models.py [hunk 3]
```diff
@@ -9,10 +9,13 @@ class Car(models.Model):
     def __unicode__(self):
         return self.plate
 
+class m2mprint:
+    def __unicode__(self):
```
Comment (nemesisdesign) [inline example/testapp/models.py]: `__unicode__` is deprecated (in python 3 there's only unicode strings). Use `__str__` with [django.utils.encoding.python_2_unicode_compatible](https://docs.djangoproject.com/en/1.10/ref/utils/#django.utils.encoding.python_2_unicode_compatible).

Subclass a real django model.

File: sortedm2m/fields.py [hunk 4]
```diff
@@ -212,17 +212,23 @@ class SortedManyToManyField(_ManyToManyField):
     '''
     Providing a many to many relation that remembers the order of related
     objects.
-
+
```
Comment (nemesisdesign) [inline sortedm2m/fields.py]: you are adding spaces and changing a line that doesn't need to be changed, try to revert this change (insert a blank line without space, commit and then squash/fixup).

File: sortedm2m/fields.py [hunk 5]
```diff
@@ -376,17 +382,19 @@ def get_intermediate_model_to_field(self, klass):
 
     def create_intermediate_model_from_attrs(self, klass, attrs):
         name = self.get_intermediate_model_name(klass)
-        return type(str(name), (models.Model,), attrs)
+        if self.base_class:
+            return type(str(name), (models.Model, self.base_class), attrs)
+        else:
+            return type(str(name), (models.Model,), attrs)
```
Comment (nemesisdesign) [inline sortedm2m/fields.py]: this whole if block can be improved. You are repeating `str(name)`, `(models.Model,)` and `attrs` twice.

You could store that tuple in an auxiliary variable and insert the base class if needed.

Could you try this? You could also try a different idea if you like.

File: sortedm2m/fields.py [hunk 6]
```diff
@@ -376,17 +382,19 @@ def get_intermediate_model_to_field(self, klass):
...
 
     def create_intermediate_model(self, klass):
         # Construct and return the new class.
         from_field_name, from_field = self.get_intermediate_model_from_field(klass)
         to_field_name, to_field = self.get_intermediate_model_to_field(klass)
         sort_value_field_name, sort_value_field = self.get_intermediate_model_sort_value_field(klass)
-
         meta = self.get_intermediate_model_meta_class(
```
Comment (nemesisdesign) [inline sortedm2m/fields.py]: more readable would be:

```python
meta = self.get_intermediate_model_meta_class(klass,
                                              from_field_name,
                                              to_field_name,
                                              sort_value_field_name)
```

File: sortedm2m/fields.py [hunk 7]
```diff
@@ -376,17 +382,19 @@ def get_intermediate_model_to_field(self, klass):
...
             klass, from_field_name, to_field_name, sort_value_field_name)
-
+
```
Comment (nemesisdesign) [inline sortedm2m/fields.py]: you are adding spaces and changing a line that doesn't need to be changed, try to revert this change (insert a blank line without space, commit and then squash/fixup).

File: example/testapp/models.py (hunk 3 above)
Comment (rohithasrk) [inline example/testapp/models.py]: `__unicode__` method is being used almost everywhere in this package. Shall I replace all the occurrences?

**Comment by rohithasrk** [comment]
Thanks for the review @nemesisdesign. I'll update the PR soon

File: example/testapp/models.py (hunk 1 above)
Comment (rohithasrk) [inline example/testapp/models.py]: Shall I use tools like [autopep8](https://pypi.python.org/pypi/autopep8) to correct all pep8 errors?

File: sortedm2m/fields.py (hunk 5 above)
Comment (rohithasrk) [inline sortedm2m/fields.py]: Yes. Seems good.

File: sortedm2m/fields.py (hunk 6 above)
Comment (rohithasrk) [inline sortedm2m/fields.py]: Yes. I agree

File: sortedm2m/fields.py (hunk 7 above)
Comment (rohithasrk) [inline sortedm2m/fields.py]: My bad. Will fix them

File: example/testapp/models.py (hunk 3 above)
Comment (nemesisdesign) [inline example/testapp/models.py]: @rohithasrk I'm looking now.. wait a minute and I'll come back to you regarding this

**Comment by nemesisdesign** [comment]
Regarding the tests, I suggest to add an assertion here:

https://github.com/gregmuellegger/django-sortedm2m/blob/master/sortedm2m_tests/test_field.py

This means you should probably move your new model to this file:

https://github.com/gregmuellegger/django-sortedm2m/blob/master/sortedm2m_tests/models.py

I will copy this text in the review text.

Federico

File: example/testapp/models.py (hunk 3 above)
Comment (nemesisdesign) [inline example/testapp/models.py]: I suggest you to move this new model to:
https://github.com/gregmuellegger/django-sortedm2m/blob/master/sortedm2m_tests/models.py

It will be easier for you to import that model in the test suite.

In the test suite you will have to ensure the representation of the model is what you expect and I'm pretty confident declaring __unicode__ only would not work on python3.
I leave the solution up to you, as long as the test works, I'm fine with it.

I pointed out that solution because in the OpenWISP modules I try to use `python_2_unicode_compatible` as much as possible but this package was probably developed before this utility was added.

File: sortedm2m/fields.py (hunk 6 above)
Comment (rohithasrk) [inline sortedm2m/fields.py]: Using tools like [autopep8](https://pypi.python.org/pypi/autopep8) should help doing all these. Anyways, will fix them soon.

File: example/testapp/models.py (hunk 1 above)
Comment (nemesisdesign) [inline example/testapp/models.py]: In [OpenWISP](http://openwisp.org) we use [flake8](https://pypi.python.org/pypi/flake8) and it is enforced in the build, but not sure is enforced here. For the moment I suggest you to focus on fixing these small issues I pointed out, it will be easier for you.

Comment (rohithasrk) [inline example/testapp/models.py]: Okay. Thanks :)

**Comment by rohithasrk** [comment]
@nemesisdesign : I've updated the PR by squashing the commits and modifying the commits with the changes you've suggested. Please review. Thanks

**Comment by nemesisdesign** [comment]
@rohithasrk great. There's only one thing missing, an assertion in the test suite. Did you get [one of my last replies](https://github.com/gregmuellegger/django-sortedm2m/pull/101#issuecomment-288027487)?

I see that you have created a model in the testapp but that is not enough. We should ensure that the test model is behaving as expected.

**Comment by nemesisdesign** [comment]
@gregmuellegger I think this PR is going to be ready soon

**Comment by rohithasrk** [comment]
@nemesisdesign : Added a test case too. I'm kinda new to writing tests, so please review. Thank you

**Comment by rohithasrk** [comment]
This test seems to be giving wrong results. I'll update the revised version in a while

**Comment by rohithasrk** [comment]
@nemesisdesign : I think this is fine. I've written a test comparing the `__unicode__` methods and not the outputs of instances. This just checks if the the `__unicode__` methods of the `m2mprint` and `shelf_books` is the same.

**Comment by rohithasrk** [comment]
@nemesisdesign @gregmuellegger I'm curious. Can this be merged?

File: sortedm2m_tests/models.py [hunk 8]
```diff
@@ -3,8 +3,14 @@
 from sortedm2m.fields import SortedManyToManyField
 
 
+class m2mprint:
+
+    def __unicode__(self):
+        return unicode(self.book)
```
Comment (nemesisdesign) [inline sortedm2m_tests/models.py]: Please do the following:

```python
class BaseCarThrough(models.Model):
    def __str__(self):
        return "Relationship to {0}".format(self.book.name)

    def __unicode__(self):
        return u"{0}".format(str(self))
```

File: sortedm2m_tests/models.py [hunk 9]
```diff
@@ -3,8 +3,14 @@
...
+
+
 class Shelf(models.Model):
-    books = SortedManyToManyField('Book', related_name='shelves')
+    books = SortedManyToManyField('Book', related_name='shelves', base_class=m2mprint)
```
Comment (nemesisdesign) [inline sortedm2m_tests/models.py]: change `m2mprint` to `BaseCarThrough`

File: example/testapp/models.py [hunk 10]
```diff
@@ -10,9 +10,15 @@ def __unicode__(self):
         return self.plate
 
 
+class m2mprint:
```
Comment (nemesisdesign) [inline example/testapp/models.py]: Get rid of all the changes to this file because they are not needed (you just moved this base model to the test suite).

File: sortedm2m_tests/test_field.py [hunk 11]
```diff
@@ -183,6 +183,17 @@ def get_ids(queryset):
...
         self.assertEqual(get_ids(shelf.books.all()), get_ids(books))
 
+    def test_base_class_m2mprint(self):
+        shelf = self.model.objects.create()
+        shelf.books.add(self.books[0])
+
+        shelf_book = shelf.books.through
+
+        if hasattr(shelf_book, '__unicode__'):
+            self.assertEqual(shelf_book.__unicode__, m2mprint.__unicode__)
```
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: @rohithasrk, this is not good enough, you should do something like:

```python
self.assertEqual(str(shelf), ''Relationship to {0}".format(shelf.name)
```

get rid of `shelf_book = shelf.books.through`.

File: sortedm2m_tests/test_field.py [hunk 12]
```diff
@@ -183,6 +183,17 @@ def get_ids(queryset):
...
+        else:
+            self.assertTrue(True)
```
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: Get rid of the if/else condition please.

BTW: adding stuff like `self.assertTrue(True)` is really bad. Don't do it!

Comment (rohithasrk) [inline sortedm2m_tests/test_field.py]: @nemesisdesign, I added these if/else conditions just to check if the base_class parameter has been provided. Tests fail when base_class parameter isn't provided. It shouldn't be the case right?

File: sortedm2m_tests/test_field.py (hunk 11 above)
Comment (rohithasrk) [inline sortedm2m_tests/test_field.py]: Shouldn't it be shelf.book.name? As shelf doesn't have an attribute name?

Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: @rohithasrk Yes

File: sortedm2m_tests/test_field.py (hunk 12 above)
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: I have not understood your last comment very well.
You have to write a test which checks the output of the `__str__` method is what you expect. No need to add if/else blocks.

File: example/testapp/models.py (hunk 10 above)
Comment (nemesisdesign) [inline example/testapp/models.py]: apply the same suggestions I gave you in https://github.com/gregmuellegger/django-sortedm2m/pull/101/files#r108227092

File: sortedm2m_tests/test_field.py (hunk 11 above)
Comment (rohithasrk) [inline sortedm2m_tests/test_field.py]: @nemesisdesign : `shelf.book.name` throws errors too, as shelf has no attribute book. That is why I've used `shelf_book = shelf.books.through`.

File: sortedm2m_tests/test_field.py (hunk 12 above)
Comment (rohithasrk) [inline sortedm2m_tests/test_field.py]: @nemesisdesign : I mean, should the test fail if a sortedm2m relationship doesn't have a base_class parameter? In my opinion, No.

As in the above mentioned changes, the test is failing. And that is why, I've used if/else conditions.

File: sortedm2m_tests/models.py (hunk 8 above)
Comment (rohithasrk) [inline sortedm2m_tests/models.py]: This throws and error saying the maximum recursion depth has exceeded. Which IMO, is maybe because `str(self)` is not invoking the `__str__` method. It is perhaps calling the `__unicode__` method. Idk, if this is the problem with python version or something. @nemesisdesign

Comment (rohithasrk) [inline sortedm2m_tests/models.py]: So, it's better to write them independently IMO. What say @nemesisdesign?

File: sortedm2m_tests/test_field.py (hunk 11 above)
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: apply the suggestions I suggested to you on IRC. 

Do something like:

```python
relationship = shelf_book.objects.first()
self.assertEqual(str(relationship), "<expected representation of the relationship here>")
```

File: sortedm2m_tests/test_field.py (hunk 12 above)
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: why should the test fail if you defined it in the model? Am I missing something?

File: sortedm2m_tests/models.py (hunk 8 above)
Comment (nemesisdesign) [inline sortedm2m_tests/models.py]: yes whatever it works as long as there is an assertion in the test that checks for the result is what we expect and the assertion works with python2 and python3.

File: sortedm2m_tests/test_field.py [hunk 13]
```diff
@@ -183,6 +183,16 @@ def get_ids(queryset):
             return [obj.id for obj in queryset]
         self.assertEqual(get_ids(shelf.books.all()), get_ids(books))
 
+    def test_base_class_str(self):
+        shelf = self.model.objects.create()
+        shelf.books.add(self.books[0])
+
```
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: remove blank linke

Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: remove blank line

File: sortedm2m_tests/test_field.py [hunk 14]
```diff
@@ -183,6 +183,16 @@ def get_ids(queryset):
...
+        through_model = shelf.books.through
+        instance = through_model.objects.first()
+
```
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: remove blank line

File: example/testapp/models.py [hunk 15]
```diff
@@ -10,9 +10,15 @@ def __unicode__(self):
         return self.plate
 
 
+class BaseCarThrough(object):
+
```
Comment (nemesisdesign) [inline example/testapp/models.py]: remove blank line

File: sortedm2m_tests/test_field.py [hunk 16]
```diff
@@ -183,6 +183,16 @@ def get_ids(queryset):
             return [obj.id for obj in queryset]
         self.assertEqual(get_ids(shelf.books.all()), get_ids(books))
 
+    def test_base_class_str(self):
+        shelf = self.model.objects.create()
+        shelf.books.add(self.books[0])
+
+        through_model = shelf.books.through
+        instance = through_model.objects.first()
```
Comment (nemesisdesign) [inline sortedm2m_tests/test_field.py]: unfortunately I forgot `first()` is not supported on old django versions.

Try rewriting this line as:

```python
instance = through_model.objects.all()[0]
```

**Comment by rohithasrk** [comment]
@gregmuellegger Resolved the conflicts are pushed. I think this can be merged.

**Comment by gregmuellegger** [comment]
I'm deeply sorry for this contribution to being so long on the waiting! I was deeply dived into personal projects (like getting maried 💒) which was that I spent nearly no time in front of a computer.

Thank you for your patience and for the contribution and the amazing guidance that you @nemesisdesign gave here!

I will release this feature in a new release in the next days.

**Comment by gregmuellegger** [comment]
It is released! You can find your contribution as part of: https://pypi.python.org/pypi/django-sortedm2m

**Comment by rohithasrk** [comment]
Thank you @gregmuellegger

**Comment by nemesisdesign** [comment]
Thanks for merging and congratulations for your marriage!
